src/core/matchers/matchersUtil.js

```diff
@@ -168,25 +205,30 @@ getJasmineRequireObj().matchersUtil = function(j$) {
     if (className == '[object Array]') {
       size = a.length;
       if (size !== b.length) {
+        diffBuilder.record(a, b);
         return false;
       }
 
-      while (size--) {
-        result = eq(a[size], b[size], aStack, bStack, customTesters);
-        if (!result) {
-          return false;
-        }
+      for (i = 0; i < size; i++) {
+        diffBuilder.withPath(i, function() {
+          result = eq(a[i], b[i], aStack, bStack, customTesters, diffBuilder) && result;
+        });
+      }
+      if (!result) {
+        return false;
       }
     } else if (className == '[object Set]') {
       if (a.size != b.size) {
+        diffBuilder.record(a, b);
         return false;
       }
       var iterA = a.values(), iterB = b.values();
       var valA, valB;
       do {
         valA = iterA.next();
         valB = iterB.next();
-        if (!eq(valA.value, valB.value, aStack, bStack, customTesters)) {
+        if (!eq(valA.value, valB.value, aStack, bStack, customTesters, j$.NullDiffBuilder())) {
```

Comment:
Sending in the NullDiffBuilder will cause both Sets to be pretty-printed in the failure message of their containing object. Basically, I was uncertain about how to display set diffs, so I just punted on this one.

Ideally, I'd like something similar to the diff that's displayed for objects when `a` and `b` have different sets of keys. However, the way set equality is implemented right now, the insertion order of items in the set matters. Given that that's the case, we could treat sets like arrays. But that, too, presents problems, because then you'd get output like

```
Expected $[0] to equal 'bar', but got 'baz'
```

when the objects in question are Sets, not Arrays. This output seems a bit misleading since set items can't actually be accessed by index.